interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,319 @@
+    DataInputStream serverIn = new DataInputStream(serverSock.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    workers.add(new Thread(new ClientReader(clientIn)));
+    workers.add(new Thread(new ClientWriter(clientOut)));
+    workers.add(new Thread(new ServerReader(serverIn)));
+    workers.add(new Thread(new ServerWriter(serverOut)));
+
+    for (Thread worker : workers) {
+      worker.start();
+    }
+
+  }
+
+  private synchronized void incrementServerQueueSize(int delta) {
+    serverQueueSize += delta;
+    if (serverQueueSize < queueLength) {
+      notifyAll();
+    }
+  }
+
+  private synchronized void incrementClientQueueSize(int delta) {
+    clientQueueSize += delta;
+    if (clientQueueSize < queueLength) {
+      notifyAll();
+    }
+  }
+
+  private synchronized int getServerChunk() {
+    while (serverQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        return 0;
+      }
+    }
+    return Math.max(0, Math.min(chunkSize, queueLength - serverQueueSize));
+  }
+
+  private synchronized int getClientChunk() {
+    while (clientQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        return 0;
+      }
+    }
+    return Math.max(0, Math.min(chunkSize, queueLength - clientQueueSize));
+  }
+
+  public static void main(String[] args) throws IOException {
+    TrafficControlProxy p = new TrafficControlProxy();
+    p.start();
+  }
+
+  private final class ClientReader implements Runnable {
+
+    private final DataInputStream clientIn;
+
+    ClientReader(DataInputStream clientIn) {
+      this.clientIn = checkNotNull(clientIn);
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = getClientChunk();
+          if (nextChunk != 0) {
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = clientIn.read(request);
+            long sendTime = System.nanoTime() + latency;
+            clientQueue.offer(new Message(sendTime, request, readableRequestBytes));
+            incrementClientQueueSize(readableRequestBytes);
+          }
+        } catch (IOException e) {
+          shutDown = true;
+        }
+      }
+    }
+  }
+
+  private final class ServerReader implements Runnable {
+
+    private final DataInputStream serverIn;
+
+    ServerReader(DataInputStream serverIn) {
+      this.serverIn = checkNotNull(serverIn);
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = getServerChunk();
+          if (nextChunk != 0) {
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = serverIn.read(request);
+            long sendTime = System.nanoTime() + latency;
+            serverQueue.offer(new Message(sendTime, request, readableRequestBytes));
```

Comment:
Swapped to put, which should always succeed on a delay queue because delay queues are unbounded.